File: python/backtracking/combinationSum.py
class Solution:
    # @param candidates, a list of integers
    # @param target, integer
    # @return a list of lists of integers
    def __init__(self):
        self.result = []
        
    def combinationSum(self, candidates, target):
        candidates.sort()
        self.dfs(0, target, [], candidates)
        return self.result
        
    def dfs(self, start, target, valuelist, candidates):
        if target == 0 and valuelist not in self.result:
            self.result.append(valuelist)
            return
        if target < 0:
            return
        for i in xrange(start, len(candidates)):
            self.dfs(i, target-candidates[i], valuelist+[candidates[i]], candidates)
if __name__ == "__main__":
    solution = Solution()

    print solution.combinationSum([2, 3, 6, 7], 7)

File: java/CombinationSum.java
public class Solution {
    public List<List<Integer>> result = new ArrayList<List<Integer>>();
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<Integer> valuelist = new ArrayList<Integer>();
        this.dfs(candidates, valuelist, target, 0);
        return this.result;
    }
    
    public void dfs(int[] candidates, List<Integer> valuelist, int target, int start) {
        if (target < 0) {
            return;
        }
        if (target == 0) {
            this.result.add(valuelist);
            return;
        }
        for (int i = start; i < candidates.length; i++) {
            List<Integer> newList = new ArrayList<Integer>(valuelist);
            newList.add(candidates[i]);
            this.dfs(candidates, newList, target-candidates[i], i);
        }
    }
}
